File: packages/sp_shared/lib/src/resolvers/category_resolver.dart
import '../models/episode_data.dart';
import '../models/smart_playlist.dart';
import '../models/smart_playlist_definition.dart';
import '../models/smart_playlist_group_def.dart';
import '../models/smart_playlist_sort.dart';
import 'smart_playlist_resolver.dart';

/// Resolver that groups episodes into predefined categories
/// by title pattern.
///
/// Reads group definitions from the definition's [groups] field.
/// Each group has a regex pattern, display name, and sort key.
/// Episodes are matched against groups in order (first match wins).
/// Groups without a pattern act as catch-all fallbacks.
class CategoryResolver implements SmartPlaylistResolver {
  @override
  String get type => 'category';

  @override
  SmartPlaylistSortRule get defaultSort => const SmartPlaylistSortRule(
    field: SmartPlaylistSortField.playlistNumber,
    order: SortOrder.ascending,
  );

  @override
  SmartPlaylistGrouping? resolve(
    List<EpisodeData> episodes,
    SmartPlaylistDefinition? definition,
  ) {
    if (definition == null) return null;

    final groupDefs = definition.groups;
    if (groupDefs == null || groupDefs.isEmpty) return null;

    return _resolveWithGroups(episodes, groupDefs);
  }

  SmartPlaylistGrouping? _resolveWithGroups(
    List<EpisodeData> episodes,
    List<SmartPlaylistGroupDef> groupDefs,
  ) {
    final patternGroups = _buildPatternGroups(groupDefs);
    String? fallbackId;
    String? fallbackDisplayName;
    bool? fallbackShowYearHeaders;

    for (final g in groupDefs) {
      if (g.pattern == null) {
        fallbackId = g.id;
        fallbackDisplayName = g.displayName;
        fallbackShowYearHeaders = g.episodeList?.showYearHeaders;
        break;
      }
    }

    final grouped = <String, List<int>>{};
    final fallbackIds = <int>[];
    final ungrouped = <int>[];

    for (final episode in episodes) {
      var matched = false;
      for (final pg in patternGroups) {
        if (pg.regex.hasMatch(episode.title)) {
          grouped.putIfAbsent(pg.id, () => []).add(episode.id);
          matched = true;
          break;
        }
      }
      if (!matched) {
        if (fallbackId != null) {
          fallbackIds.add(episode.id);
        } else {
          ungrouped.add(episode.id);
        }
      }
    }

    final groups = _buildGroups(patternGroups, grouped);
    if (fallbackIds.isNotEmpty) {
      groups.add(
        SmartPlaylistGroup(
          id: fallbackId!,
          displayName: fallbackDisplayName!,
          sortKey: groups.length + 1,
          episodeIds: fallbackIds,
          showYearHeaders: fallbackShowYearHeaders,
        ),
      );
    }

    if (groups.isEmpty && ungrouped.isEmpty) return null;

    final playlists = groups
        .map(
          (g) => SmartPlaylist(
            id: g.id,
            displayName: g.displayName,
            sortKey: g.sortKey,
            episodeIds: g.episodeIds,
          ),
        )
        .toList();

    return SmartPlaylistGrouping(
      playlists: playlists,
      ungroupedEpisodeIds: ungrouped,
      resolverType: type,
    );
  }

  List<_PatternGroup> _buildPatternGroups(
    List<SmartPlaylistGroupDef> groupDefs,
  ) {
    return groupDefs
        .where((g) => g.pattern != null)
        .map(
          (g) => _PatternGroup(
            regex: RegExp(g.pattern!),
            id: g.id,
            displayName: g.displayName,
            showYearHeaders: g.episodeList?.showYearHeaders,
          ),
        )
        .toList();
  }

  List<SmartPlaylistGroup> _buildGroups(
    List<_PatternGroup> patternGroups,
    Map<String, List<int>> grouped,
  ) {
    final groups = <SmartPlaylistGroup>[];
    var sortKey = 1;
    for (final pg in patternGroups) {
      final ids = grouped[pg.id];
      if (ids != null && ids.isNotEmpty) {
        groups.add(
          SmartPlaylistGroup(
            id: pg.id,
            displayName: pg.displayName,
            sortKey: sortKey,
            episodeIds: ids,
            showYearHeaders: pg.showYearHeaders,
          ),
        );
        sortKey++;
      }
    }
    return groups;
  }

  static PlaylistStructure parsePlaylistStructure(String? value) {
    return switch (value) {
      'grouped' => PlaylistStructure.grouped,
      _ => PlaylistStructure.split,
    };
  }

  static YearBinding parseYearBinding(String? value) {
    return switch (value) {
      'pinToYear' => YearBinding.pinToYear,
      'splitByYear' => YearBinding.splitByYear,
      _ => YearBinding.none,
    };
  }
}

class _PatternGroup {
  const _PatternGroup({
    required this.regex,
    required this.id,
    required this.displayName,
    this.showYearHeaders,
  });

  final RegExp regex;
  final String id;
  final String displayName;
  final bool? showYearHeaders;
}

File: crates/sp_core/src/resolvers/category_resolver.rs
use regex::Regex;

use crate::models::{
    EpisodeData, GroupDef, Grouping, Playlist, PlaylistDefinition, SortField, SortOrder, SortRule,
};

use super::resolver::Resolver;

/// Resolver that groups episodes into predefined categories by title pattern.
///
/// Reads group definitions from the definition's `groups` field.
/// Each group has a regex pattern, display name, and sort key.
/// Episodes are matched against groups in order (first match wins).
/// Groups without a pattern act as catch-all fallbacks.
pub struct CategoryResolver;

struct PatternGroup {
    regex: Regex,
    id: String,
    display_name: String,
    show_year_headers: Option<bool>,
}

impl Resolver for CategoryResolver {
    fn resolver_type(&self) -> &str {
        "titleClassifier"
    }

    fn default_sort(&self) -> SortRule {
        SortRule {
            field: SortField::PlaylistNumber,
            order: SortOrder::Ascending,
        }
    }

    fn resolve(
        &self,
        episodes: &[&dyn EpisodeData],
        definition: Option<&PlaylistDefinition>,
    ) -> Option<Grouping> {
        let definition = definition?;
        let group_defs = definition.effective_static_classifiers()?;
        if group_defs.is_empty() {
            return None;
        }

        resolve_with_groups(episodes, group_defs, self.resolver_type())
    }
}

fn resolve_with_groups(
    episodes: &[&dyn EpisodeData],
    group_defs: &[GroupDef],
    resolver_type: &str,
) -> Option<Grouping> {
    let pattern_groups = build_pattern_groups(group_defs);

    // Find the first fallback group (no pattern)
    let fallback = group_defs.iter().find(|g| g.pattern.is_none());
    let fallback_id = fallback.map(|g| g.id.as_str());
    let fallback_display_name = fallback.map(|g| g.display_name.as_str());
    let fallback_show_year_headers = fallback
        .and_then(|g| g.episode_list.as_ref())
        .and_then(|el| el.show_year_headers);

    // Map from pattern group id -> list of episode ids
    let mut grouped: std::collections::HashMap<String, Vec<i64>> = std::collections::HashMap::new();
    let mut fallback_ids: Vec<i64> = Vec::new();
    let mut ungrouped: Vec<i64> = Vec::new();

    for &episode in episodes {
        let mut matched = false;
        for pg in &pattern_groups {
            if pg.regex.is_match(episode.title()) {
                grouped
                    .entry(pg.id.clone())
                    .or_default()
                    .push(episode.id());
                matched = true;
                break;
            }
        }
        if !matched {
            if fallback_id.is_some() {
                fallback_ids.push(episode.id());
            } else {
                ungrouped.push(episode.id());
            }
        }
    }

    // Build playlists in definition order, assigning sequential sortKeys
    // to non-empty groups only
    let mut playlists: Vec<Playlist> = Vec::new();
    let mut sort_key = 1;

    for pg in &pattern_groups {
        if let Some(ids) = grouped.get(&pg.id)
            && !ids.is_empty()
        {
            let mut playlist = Playlist::new(
                pg.id.clone(),
                pg.display_name.clone(),
                sort_key,
                ids.clone(),
            );
            playlist.show_year_headers = pg.show_year_headers.unwrap_or(false);
            playlists.push(playlist);
            sort_key += 1;
        }
    }

    // Add fallback group last
    if !fallback_ids.is_empty() {
        let mut playlist = Playlist::new(
            fallback_id.unwrap().to_string(),
            fallback_display_name.unwrap().to_string(),
            playlists.len() as i32 + 1,
            fallback_ids,
        );
        playlist.show_year_headers = fallback_show_year_headers.unwrap_or(false);
        playlists.push(playlist);
    }

    if playlists.is_empty() {
        return None;
    }

    Some(Grouping {
        playlists,
        ungrouped_episode_ids: ungrouped,
        resolver_type: resolver_type.to_string(),
    })
}

fn build_pattern_groups(group_defs: &[GroupDef]) -> Vec<PatternGroup> {
    group_defs
        .iter()
        .filter_map(|g| {
            let pattern = g.pattern.as_ref()?;
            let regex = Regex::new(pattern).ok()?;
            Some(PatternGroup {
                regex,
                id: g.id.clone(),
                display_name: g.display_name.clone(),
                show_year_headers: g.episode_list.as_ref().and_then(|el| el.show_year_headers),
            })
        })
        .collect()
}
